# Eval Machine Learning - Catégorisation musicale 

23/04/2021 - Killian VISINE & Raphael DESBATS

## Récupération des données test et train

In [1]:
import pandas as pd

url_train = "https://raw.githubusercontent.com/RTheophile/td_ml_ynov/main/data/train.csv"
url_test = "https://raw.githubusercontent.com/RTheophile/td_ml_ynov/main/data/test.csv"

df_train = pd.read_csv(url_train, sep=',', decimal='.',index_col =0)
df_test = pd.read_csv(url_test, sep=',', decimal='.', index_col =0)

# Phase d'entrainement

#### On index "music_id" en colonne 0 <br> Puis on analyse les données avec 

*   .head() qui permet d'afficher les premiers éléments d'un dataframe
*   .info() qui décrit le contenu d'un dataframe (liste)
*   .describe() qui décrit le contenu d'un dataframe (tableaux)



In [2]:
df_train.head()

chroma_0_mean  chroma_0_std  chroma_10_mean  chroma_10_std  \
music_id                                                               
0            866.507629   1585.437134      415.300842     710.790771   
1            832.956909   2140.568115      287.584564     281.795380   
2            331.544586    553.657532      260.467499     327.618225   
3            454.521851   1170.188110      601.021790    1035.021240   
4            180.331635    245.818512      388.924744     917.082581   

          chroma_11_mean  chroma_11_std  chroma_1_mean  chroma_1_std  \
music_id                                                               
0             683.865845    1011.552856     473.484100    744.247925   
1             405.467224     754.076904    1061.359863   2263.894531   
2             283.580139     312.385986     204.369690    225.922531   
3             721.523865    1861.131836     350.991791    495.942383   
4             186.856262     192.765305     212.240402    370.108063   

          chroma_2_mean  chroma_2_std  ...  mfcc_9_std  onset_rate  \
music_id                               ...                           
0            304.914825    363.313141  ...    0.647313    3.631719   
1           2502.509033   6418.287109  ...    0.575588    6.230564   
2            286.470215    396.768219  ...    0.837297    4.264771   
3            432.551971    563.184021  ...    0.702976    4.064860   
4            170.875610    225.814850  ...    0.487534    4.164816   

          spectral_centroid_mean  spectral_decrease_mean  spectral_flux_mean  \
music_id                                                                       
0                      27.621587               -0.186330            1.962404   
1                      38.780628               -0.013198            2.139456   
2                      86.371120               -0.155013            1.336048   
3                      16.844325               -0.287745            2.207420   
4                       5.764473               -0.626706            1.263404   

          spectral_rolloff_mean  spectral_spread_mean    zcr_mean     zcr_std  \
music_id                                                                        
0                   1525.877193           1579.021838  190.907164  149.482276   
1                   1740.789474           1941.791036  257.913214  171.257592   
2                   5243.859649          15422.032531  508.122337  296.141124   
3                    988.596491            999.051613  144.638374  142.303996   
4                     21.491228            735.051431  181.818928  106.604597   

          category  
music_id            
0                1  
1                3  
2                3  
3                1  
4                1  

[5 rows x 62 columns]

In [3]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 320 entries, 0 to 319
Data columns (total 62 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   chroma_0_mean           317 non-null    float64
 1   chroma_0_std            317 non-null    float64
 2   chroma_10_mean          317 non-null    float64
 3   chroma_10_std           317 non-null    float64
 4   chroma_11_mean          318 non-null    float64
 5   chroma_11_std           320 non-null    float64
 6   chroma_1_mean           318 non-null    float64
 7   chroma_1_std            319 non-null    float64
 8   chroma_2_mean           316 non-null    float64
 9   chroma_2_std            317 non-null    float64
 10  chroma_3_mean           318 non-null    float64
 11  chroma_3_std            313 non-null    float64
 12  chroma_4_mean           316 non-null    float64
 13  chroma_4_std            316 non-null    float64
 14  chroma_5_mean           319 non-null    fl

In [4]:
df_train.describe()


chroma_0_mean  chroma_0_std  chroma_10_mean  chroma_10_std  \
count     317.000000    317.000000      317.000000     317.000000   
mean      461.223799    727.558402      441.424764     664.515463   
std       302.890638    511.932682      286.127709     507.650090   
min        52.542076    108.190132       66.789970     129.553772   
25%       254.774170    388.372406      248.708359     350.873505   
50%       377.393036    571.490845      381.081177     534.109131   
75%       592.239685    910.419189      558.673584     803.034241   
max      1643.472900   3421.152100     2172.719482    4360.215332   

       chroma_11_mean  chroma_11_std  chroma_1_mean  chroma_1_std  \
count      318.000000     320.000000     318.000000    319.000000   
mean       420.808596     606.042905     417.116685    605.835885   
std        284.887311     443.484753     269.444492    410.002545   
min         64.062683     117.359634      50.826832     93.769897   
25%        231.128601     344.294373     228.390095    353.000809   
50%        340.970337     484.871140     354.934341    505.056274   
75%        506.674622     752.747421     507.147308    702.858398   
max       2302.835205    5054.364258    1829.826904   3210.021484   

       chroma_2_mean  chroma_2_std  ...  mfcc_9_std  onset_rate  \
count     316.000000    317.000000  ...  317.000000  319.000000   
mean      469.942682    752.094412  ...    0.629552    4.917406   
std       319.000592    630.288150  ...    0.170709    2.537782   
min        50.655052    130.289444  ...    0.405036    0.366504   
25%       258.015579    377.173279  ...    0.515947    3.048645   
50%       403.037384    585.324097  ...    0.583850    4.731231   
75%       568.195679    910.971436  ...    0.683956    6.615188   
max      2502.509033   6418.287109  ...    1.771815   14.466579   

       spectral_centroid_mean  spectral_decrease_mean  spectral_flux_mean  \
count              318.000000              316.000000          316.000000   
mean                28.050904               -0.155241            1.737679   
std                 16.103779                0.185176            1.063592   
min                  1.656318               -0.650820            0.351351   
25%                 16.354873               -0.287891            0.958037   
50%                 24.205491               -0.086321            1.461911   
75%                 37.958948                0.002296            2.144044   
max                 86.371120                0.074635            6.510337   

       spectral_rolloff_mean  spectral_spread_mean    zcr_mean     zcr_std  \
count             320.000000            316.000000  316.000000  317.000000   
mean             1180.002741           1942.502831  192.060511  115.472340   
std               711.314460           1984.500644  107.853027   53.495612   
min                21.491228             41.709451   35.199924   32.996641   
25%               687.719298            626.473435  114.992507   78.780289   
50%              1020.833333           1274.624056  167.636659  104.338107   
75%              1552.741228           2634.940367  240.335595  142.303996   
max              5243.859649          15422.032531  579.817519  406.147341   

         category  
count  320.000000  
mean     1.500000  
std      1.119785  
min      0.000000  
25%      0.750000  
50%      1.500000  
75%      2.250000  
max      3.000000  

[8 rows x 62 columns]

In [5]:
X_train = df_train.dropna(thresh=2)
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 320 entries, 0 to 319
Data columns (total 62 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   chroma_0_mean           317 non-null    float64
 1   chroma_0_std            317 non-null    float64
 2   chroma_10_mean          317 non-null    float64
 3   chroma_10_std           317 non-null    float64
 4   chroma_11_mean          318 non-null    float64
 5   chroma_11_std           320 non-null    float64
 6   chroma_1_mean           318 non-null    float64
 7   chroma_1_std            319 non-null    float64
 8   chroma_2_mean           316 non-null    float64
 9   chroma_2_std            317 non-null    float64
 10  chroma_3_mean           318 non-null    float64
 11  chroma_3_std            313 non-null    float64
 12  chroma_4_mean           316 non-null    float64
 13  chroma_4_std            316 non-null    float64
 14  chroma_5_mean           319 non-null    fl

#### On peut voir qu'il y a des données manquantes car nous avons 320 entrées, pour cela nous utilisons 
<code>X_train[col].fillna(X_train[col].mean(),inplace =True)</code> <br> <h4>Permettant le remplacement des données manquantes par la moyenne des données, afin de remplir les paramètres.</h4>

In [6]:
for col in X_train.columns:
    X_train[col].fillna(X_train[col].mean(),inplace =True)
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 320 entries, 0 to 319
Data columns (total 62 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   chroma_0_mean           320 non-null    float64
 1   chroma_0_std            320 non-null    float64
 2   chroma_10_mean          320 non-null    float64
 3   chroma_10_std           320 non-null    float64
 4   chroma_11_mean          320 non-null    float64
 5   chroma_11_std           320 non-null    float64
 6   chroma_1_mean           320 non-null    float64
 7   chroma_1_std            320 non-null    float64
 8   chroma_2_mean           320 non-null    float64
 9   chroma_2_std            320 non-null    float64
 10  chroma_3_mean           320 non-null    float64
 11  chroma_3_std            320 non-null    float64
 12  chroma_4_mean           320 non-null    float64
 13  chroma_4_std            320 non-null    float64
 14  chroma_5_mean           320 non-null    fl

#### On normalise les données afin qu'elles soient toute comprise entre 0 à 1 -- Utilisation de MinMaxScaler<br><br>On a créé un tableau dataX qui contient tous les paramètres à normaliser et un tableau Label qui contient le paramètre "catagory" que l'on ne veux pas modifier

In [7]:
dataX = X_train[['chroma_0_mean', 'chroma_0_std', 'chroma_10_mean', 'chroma_10_std', 'chroma_11_mean', 'chroma_11_std', 'chroma_1_mean', 'chroma_1_std', 'chroma_2_mean', 'chroma_2_std', 'chroma_3_mean', 'chroma_3_std', 'chroma_4_mean', 'chroma_4_std', 'chroma_5_mean', 'chroma_5_std', 'chroma_6_mean', 'chroma_6_std', 'chroma_7_mean', 'chroma_7_std', 'chroma_8_mean', 'chroma_8_std', 'chroma_9_mean', 'chroma_9_std', 'chroma_centroid_mean', 'chroma_flux_mean', 'chroma_max', 'chroma_min', 'chroma_spread_mean', 'mfcc_10_mean', 'mfcc_10_std', 'mfcc_11_mean', 'mfcc_11_std', 'mfcc_12_mean', 'mfcc_12_std', 'mfcc_13_mean', 'mfcc_13_std', 'mfcc_2_mean', 'mfcc_2_std', 'mfcc_3_mean', 'mfcc_3_std', 'mfcc_4_mean', 'mfcc_4_std', 'mfcc_5_mean', 'mfcc_5_std', 'mfcc_6_mean', 'mfcc_6_std', 'mfcc_7_mean', 'mfcc_7_std', 'mfcc_8_mean', 'mfcc_8_std', 'mfcc_9_mean', 'mfcc_9_std', 'onset_rate', 'spectral_centroid_mean', 'spectral_decrease_mean', 'spectral_flux_mean', 'spectral_rolloff_mean', 'spectral_spread_mean', 'zcr_mean', 'zcr_std']]

Label = X_train['category']

from sklearn.preprocessing import MinMaxScaler
summaryData = dataX.describe()
DataCol=list(summaryData)

scaler = MinMaxScaler(feature_range=(0,1))
Data =scaler.fit_transform(dataX)
Data = pd.DataFrame(Data,columns=DataCol)
X_train[['chroma_0_mean', 'chroma_0_std', 'chroma_10_mean', 'chroma_10_std', 'chroma_11_mean', 'chroma_11_std', 'chroma_1_mean', 'chroma_1_std', 'chroma_2_mean', 'chroma_2_std', 'chroma_3_mean', 'chroma_3_std', 'chroma_4_mean', 'chroma_4_std', 'chroma_5_mean', 'chroma_5_std', 'chroma_6_mean', 'chroma_6_std', 'chroma_7_mean', 'chroma_7_std', 'chroma_8_mean', 'chroma_8_std', 'chroma_9_mean', 'chroma_9_std', 'chroma_centroid_mean', 'chroma_flux_mean', 'chroma_max', 'chroma_min', 'chroma_spread_mean', 'mfcc_10_mean', 'mfcc_10_std', 'mfcc_11_mean', 'mfcc_11_std', 'mfcc_12_mean', 'mfcc_12_std', 'mfcc_13_mean', 'mfcc_13_std', 'mfcc_2_mean', 'mfcc_2_std', 'mfcc_3_mean', 'mfcc_3_std', 'mfcc_4_mean', 'mfcc_4_std', 'mfcc_5_mean', 'mfcc_5_std', 'mfcc_6_mean', 'mfcc_6_std', 'mfcc_7_mean', 'mfcc_7_std', 'mfcc_8_mean', 'mfcc_8_std', 'mfcc_9_mean', 'mfcc_9_std', 'onset_rate', 'spectral_centroid_mean', 'spectral_decrease_mean', 'spectral_flux_mean', 'spectral_rolloff_mean', 'spectral_spread_mean', 'zcr_mean', 'zcr_std']]=Data

#### Data correspond au paramètres normaliser sans "category"

In [8]:
Data.describe()

chroma_0_mean  chroma_0_std  chroma_10_mean  chroma_10_std  \
count     320.000000    320.000000      320.000000     320.000000   
mean        0.256882      0.186953        0.177895       0.126449   
std         0.189488      0.153796        0.135227       0.119428   
min         0.000000      0.000000        0.000000       0.000000   
25%         0.127726      0.084711        0.086704       0.052323   
50%         0.205260      0.140712        0.149900       0.096342   
75%         0.339150      0.242128        0.231683       0.158359   
max         1.000000      1.000000        1.000000       1.000000   

       chroma_11_mean  chroma_11_std  chroma_1_mean  chroma_1_std  \
count      320.000000     320.000000     320.000000    320.000000   
mean         0.159349       0.098984       0.205896      0.164321   
std          0.126852       0.089829       0.150983      0.131363   
min          0.000000       0.000000       0.000000      0.000000   
25%          0.074796       0.045966       0.100096      0.083355   
50%          0.124880       0.074440       0.173173      0.132052   
75%          0.195893       0.128699       0.256365      0.195213   
max          1.000000       1.000000       1.000000      1.000000   

       chroma_2_mean  chroma_2_std  ...  mfcc_9_mean  mfcc_9_std  onset_rate  \
count     320.000000    320.000000  ...   320.000000  320.000000  320.000000   
mean        0.171008      0.098888  ...     0.320870    0.164267    0.322757   
std         0.129288      0.099764  ...     0.097619    0.124310    0.179701   
min         0.000000      0.000000  ...     0.000000    0.000000    0.000000   
25%         0.085413      0.039533  ...     0.265996    0.081744    0.190813   
50%         0.144277      0.073136  ...     0.322390    0.132633    0.309553   
75%         0.210068      0.123054  ...     0.370991    0.203431    0.441343   
max         1.000000      1.000000  ...     1.000000    1.000000    1.000000   

       spectral_centroid_mean  spectral_decrease_mean  spectral_flux_mean  \
count              320.000000              320.000000          320.000000   
mean                 0.311570                0.683128            0.225090   
std                  0.189497                0.253650            0.171603   
min                  0.000000                0.000000            0.000000   
25%                  0.173537                0.506255            0.099249   
50%                  0.268074                0.774655            0.181609   
75%                  0.427171                0.899583            0.289530   
max                  1.000000                1.000000            1.000000   

       spectral_rolloff_mean  spectral_spread_mean    zcr_mean     zcr_std  
count             320.000000            320.000000  320.000000  320.000000  
mean                0.221836              0.123586    0.288020    0.221025  
std                 0.136205              0.128217    0.196789    0.142686  
min                 0.000000              0.000000    0.000000    0.000000  
25%                 0.127572              0.038151    0.146856    0.122904  
50%                 0.191358              0.086122    0.249123    0.192723  
75%                 0.293210              0.166656    0.374324    0.292528  
max                 1.000000              1.000000    1.000000    1.000000  

[8 rows x 61 columns]

#### X_train correspond au paramètres normaliser avec "category"

In [9]:
X_train.describe()

chroma_0_mean  chroma_0_std  chroma_10_mean  chroma_10_std  \
count     320.000000    320.000000      320.000000     320.000000   
mean        0.256882      0.186953        0.177895       0.126449   
std         0.189488      0.153796        0.135227       0.119428   
min         0.000000      0.000000        0.000000       0.000000   
25%         0.127726      0.084711        0.086704       0.052323   
50%         0.205260      0.140712        0.149900       0.096342   
75%         0.339150      0.242128        0.231683       0.158359   
max         1.000000      1.000000        1.000000       1.000000   

       chroma_11_mean  chroma_11_std  chroma_1_mean  chroma_1_std  \
count      320.000000     320.000000     320.000000    320.000000   
mean         0.159349       0.098984       0.205896      0.164321   
std          0.126852       0.089829       0.150983      0.131363   
min          0.000000       0.000000       0.000000      0.000000   
25%          0.074796       0.045966       0.100096      0.083355   
50%          0.124880       0.074440       0.173173      0.132052   
75%          0.195893       0.128699       0.256365      0.195213   
max          1.000000       1.000000       1.000000      1.000000   

       chroma_2_mean  chroma_2_std  ...  mfcc_9_std  onset_rate  \
count     320.000000    320.000000  ...  320.000000  320.000000   
mean        0.171008      0.098888  ...    0.164267    0.322757   
std         0.129288      0.099764  ...    0.124310    0.179701   
min         0.000000      0.000000  ...    0.000000    0.000000   
25%         0.085413      0.039533  ...    0.081744    0.190813   
50%         0.144277      0.073136  ...    0.132633    0.309553   
75%         0.210068      0.123054  ...    0.203431    0.441343   
max         1.000000      1.000000  ...    1.000000    1.000000   

       spectral_centroid_mean  spectral_decrease_mean  spectral_flux_mean  \
count              320.000000              320.000000          320.000000   
mean                 0.311570                0.683128            0.225090   
std                  0.189497                0.253650            0.171603   
min                  0.000000                0.000000            0.000000   
25%                  0.173537                0.506255            0.099249   
50%                  0.268074                0.774655            0.181609   
75%                  0.427171                0.899583            0.289530   
max                  1.000000                1.000000            1.000000   

       spectral_rolloff_mean  spectral_spread_mean    zcr_mean     zcr_std  \
count             320.000000            320.000000  320.000000  320.000000   
mean                0.221836              0.123586    0.288020    0.221025   
std                 0.136205              0.128217    0.196789    0.142686   
min                 0.000000              0.000000    0.000000    0.000000   
25%                 0.127572              0.038151    0.146856    0.122904   
50%                 0.191358              0.086122    0.249123    0.192723   
75%                 0.293210              0.166656    0.374324    0.292528   
max                 1.000000              1.000000    1.000000    1.000000   

         category  
count  320.000000  
mean     1.500000  
std      1.119785  
min      0.000000  
25%      0.750000  
50%      1.500000  
75%      2.250000  
max      3.000000  

[8 rows x 62 columns]

#### On vérifie la corrélation entre tout les paramètres, afin de déterminer lesquels sont les plus importants pour que le modèle soit le plus précie possible et ainsi réduire la dimension des données à interpréter.

In [10]:
from scipy import stats
columns = ['chroma_0_mean', 'chroma_0_std', 'chroma_10_mean', 'chroma_10_std', 'chroma_11_mean', 
           'chroma_11_std', 'chroma_1_mean', 'chroma_1_std', 'chroma_2_mean', 'chroma_2_std', 'chroma_3_mean', 
           'chroma_3_std', 'chroma_4_mean', 'chroma_4_std', 'chroma_5_mean', 'chroma_5_std', 'chroma_6_mean', 
           'chroma_6_std', 'chroma_7_mean', 'chroma_7_std', 'chroma_8_mean', 'chroma_8_std', 'chroma_9_mean', 
           'chroma_9_std', 'chroma_centroid_mean', 'chroma_flux_mean', 'chroma_max', 'chroma_min', 'chroma_spread_mean', 
           'mfcc_10_mean', 'mfcc_10_std', 'mfcc_11_mean', 'mfcc_11_std', 'mfcc_12_mean', 'mfcc_12_std', 'mfcc_13_mean', 
           'mfcc_13_std', 'mfcc_2_mean', 'mfcc_2_std', 'mfcc_3_mean', 'mfcc_3_std', 'mfcc_4_mean', 'mfcc_4_std', 'mfcc_5_mean', 
           'mfcc_5_std', 'mfcc_6_mean', 'mfcc_6_std', 'mfcc_7_mean', 'mfcc_7_std', 'mfcc_8_mean', 'mfcc_8_std', 'mfcc_9_mean', 
           'mfcc_9_std', 'onset_rate', 'spectral_centroid_mean', 'spectral_decrease_mean', 'spectral_flux_mean', 
           'spectral_rolloff_mean', 'spectral_spread_mean', 'zcr_mean', 'zcr_std']
Y = Label
res = []
for j in range(len(columns)-1):
    slope, intercept, r_value, p_value, _ = stats.linregress(X_train[columns[j]],Y)
    
    print('modele ', j+1)
    print('----- Valeur Totale = ', intercept, ' + ', slope,' * ', columns[j])
    print('----- R² = ', r_value)
    print('      Donc {:2.2f}% de la variabilité de la valeur totale est expliquée par {:s}'.format(r_value *100,columns[j]))
    print('----- p value = ', p_value)
    print('      H0: beta1 = 0')
    
    if p_value < 0.05 :
        print('      ',p_value,'<=5%','==> on rejette H0, au seuil de risque alpha = 5%')
        print('          le modèle ',j+1, 'est bien pertinent.')
        res.append(columns[j])
    else : 
        print('      ',p_value,'> 5%','==> on accepte H0, au seuil de risque alpha = 5%')
        print('          le modèle ',j+1, 'n est pas pertinent car pas statistiquement significatif.')
print(res)
X = X_train[res]

len(res)

modele  1
----- Valeur Totale =  1.2998589182798355  +  0.7791163601692646  *  chroma_0_mean
----- R² =  0.1318409802890763
      Donc 13.18% de la variabilité de la valeur totale est expliquée par chroma_0_mean
----- p value =  0.018298476999744226
      H0: beta1 = 0
       0.018298476999744226 <=5% ==> on rejette H0, au seuil de risque alpha = 5%
          le modèle  1 est bien pertinent.
modele  2
----- Valeur Totale =  1.4014590930493993  +  0.5270891213346802  *  chroma_0_std
----- R² =  0.07239257152769188
      Donc 7.24% de la variabilité de la valeur totale est expliquée par chroma_0_std
----- p value =  0.1964869473982334
      H0: beta1 = 0
       0.1964869473982334 > 5% ==> on accepte H0, au seuil de risque alpha = 5%
          le modèle  2 n est pas pertinent car pas statistiquement significatif.
modele  3
----- Valeur Totale =  1.4061375723822596  +  0.5276275988979883  *  chroma_10_mean
----- R² =  0.06371726663160654
      Donc 6.37% de la variabilité de la valeur tota

33

#### Nous avons 33 paramètres déterminants

In [11]:
import numpy as np
from sklearn.decomposition import PCA
pca = PCA(n_components=33)
pca.fit(X_train)

PCA(n_components=33)

In [12]:
X_train_reduc = pca.transform(X_train)
print(X_train_reduc)

[[ 0.49331366  0.28552087  0.05959616 ...  0.11437038  0.18263764
   0.01927325]
 [-1.6343919   0.83817718  0.67216246 ... -0.18263796 -0.1108924
   0.05074204]
 [-1.50508055 -0.09860585 -0.09355767 ... -0.13282955  0.02464179
  -0.02257338]
 ...
 [ 0.51314188 -0.12798807 -0.11900175 ... -0.02393963 -0.02232482
  -0.03633264]
 [ 0.44993609  0.25992387 -0.2516951  ... -0.11348766  0.01991911
   0.10894554]
 [-0.68779521  0.95148735 -0.4585746  ...  0.07308343 -0.03900223
  -0.07340599]]


In [13]:
print(pca.explained_variance_ratio_)

[0.50053789 0.12980662 0.07641414 0.03518802 0.03476181 0.02389218
 0.01734425 0.0165512  0.01388449 0.0125723  0.01121879 0.01019689
 0.0091771  0.00850859 0.0082659  0.00733886 0.00663719 0.00644299
 0.00527258 0.00481151 0.00469799 0.00440008 0.00383379 0.00375869
 0.00349125 0.00329571 0.00297159 0.00278466 0.00263398 0.00240724
 0.00214932 0.00206449 0.00197471]


In [14]:
result = 0
for ratio in pca.explained_variance_ratio_:
    result+=ratio
print(result)

0.9792867822440949


#### Nous avons transformer les paramètres en pourcentage afin de calculer le ratio de précision, notre PCA est donc de 97,9% <br><br> Ensuite nous utilisons train_test_split afin de créer un échantillon non biaisé.

In [15]:
from sklearn.model_selection import train_test_split

X_train, X_val, y_train, y_val = train_test_split(X, Y, random_state = 0)
print(X_train, X_val, y_train, y_val)

          chroma_0_mean  chroma_1_mean  chroma_1_std  chroma_3_mean  \
music_id                                                              
198            0.426399       0.466177      0.338110       0.409477   
238            0.338580       0.255664      0.139862       0.233189   
199            0.049966       0.035209      0.023508       0.187017   
222            0.003057       0.003016      0.034921       0.044336   
120            0.106860       0.054572      0.057487       0.124769   
...                 ...            ...           ...            ...   
251            0.325919       0.239517      0.162266       0.285385   
192            0.214324       0.227348      0.185585       0.304420   
117            0.194476       0.250340      0.295802       0.265128   
47             0.866393       0.779889      0.271566       0.733043   
172            0.152555       0.091115      0.108423       0.123671   

          chroma_3_std  chroma_4_mean  chroma_4_std  chroma_5_mean  \
music_

### Modele SVM
#### Il s'agit de notre algorithme de classification

In [16]:
from sklearn.svm import SVC
from sklearn import metrics

svclassifier = SVC(kernel='linear')
svclassifier.fit(X_train, y_train)

y_pred = svclassifier.predict(X_val)

### Accuracy & Matrice de confusion
#### Performance de notre modele = 80%

In [17]:
scores = metrics.accuracy_score(y_val, y_pred)
print('Accuracy: ','{:2.2%}'.format(scores))

cm = metrics.confusion_matrix(y_val, y_pred)
print(cm)

Accuracy:  80.00%
[[16  1  0  4]
 [ 1 18  0  1]
 [ 0  3 18  0]
 [ 1  3  2 12]]


#### On a comparé notre algorithme SVM avec un KNN, dont l'accuracy est respectivement 80% contre 72,5% <br> Le modele SVM est donc plus précis

In [18]:
from sklearn.neighbors import KNeighborsClassifier
knn2 = KNeighborsClassifier()
knn2.fit(X_train, y_train)

y_pred2 = knn2.predict(X_val)

from sklearn.metrics import accuracy_score
acc = accuracy_score(y_val, y_pred2)
print('Accuracy: {:2.2%}'.format(acc))

def best_model_search(X_train,X_val,y_train,y_val, n=10, patience=3):
    k=1
    best_acc = 0
    best_model=None
    best_k=1
    while k<n and patience>0:
        knn =KNeighborsClassifier(n_neighbors = k)
        knn.fit(X_train,y_train)
        y_pred = knn.predict(X_val)
        accuracy = accuracy_score(y_val, y_pred)
        print('accuracy : {:.2%} '.format(accuracy),'avec k = ',k)
        if accuracy > best_acc:
            best_acc = accuracy
            best_model = knn
            best_k=k
        patience = patience-1
        k=k+1
    return best_model,best_k,best_acc

knn2, k, acc = best_model_search(X_train,X_val,y_train,y_val,n=100, patience=30)
print('best_model = ',knn2,' accuracy : {:.2%} '.format(acc),'Pour un k = ',k)

Accuracy: 72.50%
accuracy : 68.75%  avec k =  1
accuracy : 62.50%  avec k =  2
accuracy : 66.25%  avec k =  3
accuracy : 67.50%  avec k =  4
accuracy : 72.50%  avec k =  5
accuracy : 67.50%  avec k =  6
accuracy : 72.50%  avec k =  7
accuracy : 70.00%  avec k =  8
accuracy : 67.50%  avec k =  9
accuracy : 68.75%  avec k =  10
accuracy : 68.75%  avec k =  11
accuracy : 68.75%  avec k =  12
accuracy : 66.25%  avec k =  13
accuracy : 66.25%  avec k =  14
accuracy : 67.50%  avec k =  15
accuracy : 67.50%  avec k =  16
accuracy : 68.75%  avec k =  17
accuracy : 68.75%  avec k =  18
accuracy : 71.25%  avec k =  19
accuracy : 70.00%  avec k =  20
accuracy : 70.00%  avec k =  21
accuracy : 70.00%  avec k =  22
accuracy : 70.00%  avec k =  23
accuracy : 70.00%  avec k =  24
accuracy : 70.00%  avec k =  25
accuracy : 70.00%  avec k =  26
accuracy : 67.50%  avec k =  27
accuracy : 70.00%  avec k =  28
accuracy : 68.75%  avec k =  29
accuracy : 68.75%  avec k =  30
best_model =  KNeighborsClassifi

# Phase de test - Prédiction 
#### On re-effectue les étapes précédente : 
<ul><li>Récupération des données</li><li>Analyse des données</li><li>Remplacement des données manquantes</li><li>Normalisation des données</li><li>Prédiction des données</li><li>Création du csv</li>
    
### Analyse des données

In [19]:
df_test.head()

chroma_0_mean  chroma_0_std  chroma_10_mean  chroma_10_std  \
music_id                                                               
0            232.907440    350.553833      206.571701     331.046967   
1            342.126831    615.817993      447.988678     562.143860   
2             92.768356    232.246658      152.421188     268.086456   
3            219.882584    410.804382      199.726151     305.265564   
4            215.454269    306.737244      259.520355     343.783783   

          chroma_11_mean  chroma_11_std  chroma_1_mean  chroma_1_std  \
music_id                                                               
0             225.048828     347.441132     221.465195    332.978180   
1             413.652649     633.503174     432.197021    668.721008   
2              79.888550     135.265106      75.790894    156.143219   
3             452.057037    1210.232788     321.175476    815.782837   
4             305.393311     368.144043     238.357651    349.670929   

          chroma_2_mean  chroma_2_std  ...  mfcc_9_mean  mfcc_9_std  \
music_id                               ...                            
0            224.200668    347.038422  ...    -0.138428    0.572161   
1            359.124786    426.877014  ...     0.132142    0.684787   
2            148.015030    304.408112  ...    -0.225053    0.602654   
3            408.686401    698.333984  ...    -0.261579    0.970925   
4            203.332275    340.900879  ...    -0.130924    0.526003   

          onset_rate  spectral_centroid_mean  spectral_decrease_mean  \
music_id                                                               
0           8.366616               56.939299                0.046856   
1           4.198134               23.288761               -0.022325   
2           0.866282               20.381682               -0.263736   
3           1.965793               32.093015               -0.157930   
4           4.298090               59.155876               -0.025128   

          spectral_flux_mean  spectral_rolloff_mean  spectral_spread_mean  \
music_id                                                                    
0                   1.796432             472.807018           2692.618591   
1                   1.785995            1310.964912           1230.997277   
2                   0.476780            1031.578947           1006.524430   
3                   0.866502            1332.456140            912.472403   
4                   1.741103            1482.894737           4637.320310   

            zcr_mean     zcr_std  
music_id                          
0         163.125244  115.239277  
1         170.899800  113.006177  
2          81.420967   69.912374  
3         148.232375   65.696249  
4         225.845447  110.317563  

[5 rows x 61 columns]

In [20]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 80 entries, 0 to 79
Data columns (total 61 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   chroma_0_mean           78 non-null     float64
 1   chroma_0_std            80 non-null     float64
 2   chroma_10_mean          79 non-null     float64
 3   chroma_10_std           79 non-null     float64
 4   chroma_11_mean          78 non-null     float64
 5   chroma_11_std           79 non-null     float64
 6   chroma_1_mean           80 non-null     float64
 7   chroma_1_std            79 non-null     float64
 8   chroma_2_mean           78 non-null     float64
 9   chroma_2_std            78 non-null     float64
 10  chroma_3_mean           80 non-null     float64
 11  chroma_3_std            79 non-null     float64
 12  chroma_4_mean           80 non-null     float64
 13  chroma_4_std            79 non-null     float64
 14  chroma_5_mean           77 non-null     floa

In [21]:
df_test.describe()

chroma_0_mean  chroma_0_std  chroma_10_mean  chroma_10_std  \
count      78.000000     80.000000       79.000000      79.000000   
mean      499.019192    756.304766      443.320154     608.541475   
std       338.514721    588.368496      304.756461     407.868544   
min        53.816399    143.050186       50.539963     169.004440   
25%       288.748695    363.656563      243.882126     349.017593   
50%       381.551727    599.016907      371.654480     561.518738   
75%       620.127731    912.371628      517.678070     761.920624   
max      1445.304932   2985.768555     1540.302002    3276.458496   

       chroma_11_mean  chroma_11_std  chroma_1_mean  chroma_1_std  \
count       78.000000      79.000000      80.000000     79.000000   
mean       441.346557     579.474184     424.730119    604.296759   
std        279.389050     318.259395     281.714756    400.998751   
min         24.518885      66.645477      61.858105    143.562088   
25%        258.705696     360.859451     244.979664    353.203735   
50%        401.423737     539.162781     370.413406    535.344604   
75%        532.698334     707.809967     478.802505    732.788208   
max       1656.597778    1993.624146    1370.596436   3001.425781   

       chroma_2_mean  chroma_2_std  ...  mfcc_9_mean  mfcc_9_std  onset_rate  \
count      78.000000     78.000000  ...    79.000000   80.000000   79.000000   
mean      441.953600    598.358575  ...     0.060856    0.650914    4.968134   
std       302.778186    323.155955  ...     0.251296    0.230513    2.473411   
min        69.170036    116.803413  ...    -0.640222    0.428611    0.266548   
25%       241.198341    371.768143  ...    -0.104857    0.508067    3.115282   
50%       361.450089    503.391418  ...     0.035343    0.558695    5.097734   
75%       543.408447    767.574097  ...     0.199246    0.709743    6.530431   
max      1304.457153   1727.041016  ...     0.869054    1.549749   11.092346   

       spectral_centroid_mean  spectral_decrease_mean  spectral_flux_mean  \
count               79.000000               79.000000           80.000000   
mean                28.706961               -0.121660            1.690073   
std                 14.841098                0.161541            1.042189   
min                  4.779403               -0.580912            0.316866   
25%                 17.352620               -0.197562            0.964552   
50%                 24.821225               -0.068054            1.448118   
75%                 38.284458                0.006803            2.112810   
max                 64.008354                0.049315            6.491353   

       spectral_rolloff_mean  spectral_spread_mean    zcr_mean     zcr_std  
count              80.000000             80.000000   80.000000   80.000000  
mean             1169.660088           1874.662294  198.508091  111.645468  
std               547.139064           1670.692483  122.582350   45.884444  
min               128.947368            121.619467   40.129422   40.611415  
25%               703.837719            585.134696  119.962242   82.463138  
50%              1117.543860           1238.039693  166.928174  104.292470  
75%              1568.859649           2718.188834  227.851144  136.666511  
max              2514.473684           6828.632552  613.866728  297.225912  

[8 rows x 61 columns]

### Remplacement des données manquantes

In [22]:
X_test = df_test
for col in X_test.columns:
    X_test[col].fillna(X_test[col].mean(),inplace =True)
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 80 entries, 0 to 79
Data columns (total 61 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   chroma_0_mean           80 non-null     float64
 1   chroma_0_std            80 non-null     float64
 2   chroma_10_mean          80 non-null     float64
 3   chroma_10_std           80 non-null     float64
 4   chroma_11_mean          80 non-null     float64
 5   chroma_11_std           80 non-null     float64
 6   chroma_1_mean           80 non-null     float64
 7   chroma_1_std            80 non-null     float64
 8   chroma_2_mean           80 non-null     float64
 9   chroma_2_std            80 non-null     float64
 10  chroma_3_mean           80 non-null     float64
 11  chroma_3_std            80 non-null     float64
 12  chroma_4_mean           80 non-null     float64
 13  chroma_4_std            80 non-null     float64
 14  chroma_5_mean           80 non-null     floa

### Normalisation des données

In [23]:
summaryData = X_test.describe()
DataCol=list(summaryData)

scaler = MinMaxScaler(feature_range=(0,1))
X_test =scaler.fit_transform(X_test)
X_test = pd.DataFrame(X_test,columns=DataCol)

X_test.describe()

chroma_0_mean  chroma_0_std  chroma_10_mean  chroma_10_std  \
count      80.000000     80.000000       80.000000      80.000000   
mean        0.319947      0.215728        0.263653       0.141446   
std         0.240176      0.206974        0.203268       0.130422   
min         0.000000      0.000000        0.000000       0.000000   
25%         0.172077      0.077604        0.130241       0.058150   
50%         0.239954      0.160398        0.217472       0.126414   
75%         0.404746      0.270629        0.312585       0.190290   
max         1.000000      1.000000        1.000000       1.000000   

       chroma_11_mean  chroma_11_std  chroma_1_mean  chroma_1_std  \
count       80.000000      80.000000      80.000000     80.000000   
mean         0.255397       0.266131       0.277269      0.161216   
std          0.169005       0.164111       0.215257      0.139423   
min          0.000000       0.000000       0.000000      0.000000   
25%          0.151303       0.152727       0.139922      0.073974   
50%          0.234732       0.245904       0.235765      0.139694   
75%          0.299583       0.331429       0.318585      0.205853   
max          1.000000       1.000000       1.000000      1.000000   

       chroma_2_mean  chroma_2_std  ...  mfcc_9_mean  mfcc_9_std  onset_rate  \
count      80.000000     80.000000  ...    80.000000   80.000000   80.000000   
mean        0.301779      0.299058  ...     0.464513    0.198283    0.434295   
std         0.241985      0.198132  ...     0.165444    0.205607    0.227023   
min         0.000000      0.000000  ...     0.000000    0.000000    0.000000   
25%         0.145191      0.160850  ...     0.355879    0.070870    0.263913   
50%         0.249111      0.247379  ...     0.456061    0.116028    0.446266   
75%         0.381457      0.398618  ...     0.555698    0.250756    0.570913   
max         1.000000      1.000000  ...     1.000000    1.000000    1.000000   

       spectral_centroid_mean  spectral_decrease_mean  spectral_flux_mean  \
count               80.000000               80.000000           80.000000   
mean                 0.403984                0.728709            0.222400   
std                  0.248981                0.254694            0.168790   
min                  0.000000                0.000000            0.000000   
25%                  0.221740                0.611314            0.104897   
50%                  0.348565                0.811930            0.183214   
75%                  0.559718                0.931793            0.290865   
max                  1.000000                1.000000            1.000000   

       spectral_rolloff_mean  spectral_spread_mean   zcr_mean    zcr_std  
count              80.000000             80.000000  80.000000  80.000000  
mean                0.436261              0.261375   0.276047   0.276812  
std                 0.229358              0.249096   0.213656   0.178807  
min                 0.000000              0.000000   0.000000   0.000000  
25%                 0.240991              0.069109   0.139145   0.163092  
50%                 0.414414              0.166456   0.221005   0.248158  
75%                 0.603604              0.387142   0.327191   0.374317  
max                 1.000000              1.000000   1.000000   1.000000  

[8 rows x 61 columns]

In [24]:
X_topred = X_test[res]
X_topred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80 entries, 0 to 79
Data columns (total 33 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   chroma_0_mean           80 non-null     float64
 1   chroma_1_mean           80 non-null     float64
 2   chroma_1_std            80 non-null     float64
 3   chroma_3_mean           80 non-null     float64
 4   chroma_3_std            80 non-null     float64
 5   chroma_4_mean           80 non-null     float64
 6   chroma_4_std            80 non-null     float64
 7   chroma_5_mean           80 non-null     float64
 8   chroma_6_mean           80 non-null     float64
 9   chroma_7_mean           80 non-null     float64
 10  chroma_flux_mean        80 non-null     float64
 11  chroma_max              80 non-null     float64
 12  mfcc_10_std             80 non-null     float64
 13  mfcc_11_mean            80 non-null     float64
 14  mfcc_11_std             80 non-null     floa

### Prédiction des données

In [25]:
y_pred = svclassifier.predict(X_topred)

In [26]:
print(y_pred)
print(len(y_pred))

[2 1 0 0 2 1 2 2 1 2 0 2 2 2 1 3 0 3 3 3 1 1 1 0 3 2 3 0 3 2 3 2 0 2 1 1 0
 0 3 3 2 3 2 1 3 1 1 0 3 0 3 0 2 2 2 3 3 1 0 2 3 2 0 3 2 1 2 1 3 2 0 2 1 2
 2 1 1 3 0 0]
80


In [27]:
df_test["Prediction"] = y_pred
df_test.head()

chroma_0_mean  chroma_0_std  chroma_10_mean  chroma_10_std  \
music_id                                                               
0            232.907440    350.553833      206.571701     331.046967   
1            342.126831    615.817993      447.988678     562.143860   
2             92.768356    232.246658      152.421188     268.086456   
3            219.882584    410.804382      199.726151     305.265564   
4            215.454269    306.737244      259.520355     343.783783   

          chroma_11_mean  chroma_11_std  chroma_1_mean  chroma_1_std  \
music_id                                                               
0             225.048828     347.441132     221.465195    332.978180   
1             413.652649     633.503174     432.197021    668.721008   
2              79.888550     135.265106      75.790894    156.143219   
3             452.057037    1210.232788     321.175476    815.782837   
4             305.393311     368.144043     238.357651    349.670929   

          chroma_2_mean  chroma_2_std  ...  mfcc_9_std  onset_rate  \
music_id                               ...                           
0            224.200668    347.038422  ...    0.572161    8.366616   
1            359.124786    426.877014  ...    0.684787    4.198134   
2            148.015030    304.408112  ...    0.602654    0.866282   
3            408.686401    698.333984  ...    0.970925    1.965793   
4            203.332275    340.900879  ...    0.526003    4.298090   

          spectral_centroid_mean  spectral_decrease_mean  spectral_flux_mean  \
music_id                                                                       
0                      56.939299                0.046856            1.796432   
1                      23.288761               -0.022325            1.785995   
2                      20.381682               -0.263736            0.476780   
3                      32.093015               -0.157930            0.866502   
4                      59.155876               -0.025128            1.741103   

          spectral_rolloff_mean  spectral_spread_mean    zcr_mean     zcr_std  \
music_id                                                                        
0                    472.807018           2692.618591  163.125244  115.239277   
1                   1310.964912           1230.997277  170.899800  113.006177   
2                   1031.578947           1006.524430   81.420967   69.912374   
3                   1332.456140            912.472403  148.232375   65.696249   
4                   1482.894737           4637.320310  225.845447  110.317563   

          Prediction  
music_id              
0                  2  
1                  1  
2                  0  
3                  0  
4                  2  

[5 rows x 62 columns]

In [28]:
final=df_test.reset_index(drop=False)
final.head()

music_id  chroma_0_mean  chroma_0_std  chroma_10_mean  chroma_10_std  \
0         0     232.907440    350.553833      206.571701     331.046967   
1         1     342.126831    615.817993      447.988678     562.143860   
2         2      92.768356    232.246658      152.421188     268.086456   
3         3     219.882584    410.804382      199.726151     305.265564   
4         4     215.454269    306.737244      259.520355     343.783783   

   chroma_11_mean  chroma_11_std  chroma_1_mean  chroma_1_std  chroma_2_mean  \
0      225.048828     347.441132     221.465195    332.978180     224.200668   
1      413.652649     633.503174     432.197021    668.721008     359.124786   
2       79.888550     135.265106      75.790894    156.143219     148.015030   
3      452.057037    1210.232788     321.175476    815.782837     408.686401   
4      305.393311     368.144043     238.357651    349.670929     203.332275   

   ...  mfcc_9_std  onset_rate  spectral_centroid_mean  \
0  ...    0.572161    8.366616               56.939299   
1  ...    0.684787    4.198134               23.288761   
2  ...    0.602654    0.866282               20.381682   
3  ...    0.970925    1.965793               32.093015   
4  ...    0.526003    4.298090               59.155876   

   spectral_decrease_mean  spectral_flux_mean  spectral_rolloff_mean  \
0                0.046856            1.796432             472.807018   
1               -0.022325            1.785995            1310.964912   
2               -0.263736            0.476780            1031.578947   
3               -0.157930            0.866502            1332.456140   
4               -0.025128            1.741103            1482.894737   

   spectral_spread_mean    zcr_mean     zcr_std  Prediction  
0           2692.618591  163.125244  115.239277           2  
1           1230.997277  170.899800  113.006177           1  
2           1006.524430   81.420967   69.912374           0  
3            912.472403  148.232375   65.696249           0  
4           4637.320310  225.845447  110.317563           2  

[5 rows x 63 columns]

### Création du fichier CSV

In [29]:
final[["music_id","Prediction"]].to_csv('test.csv' , sep=',', decimal='.', index=False)

#### <center>Fait par Killian VISINE & Raphael DESBATS, le 23/04/2021